# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/ipc/CanvasParent.h
# Commit: 4357d695b8d5
# Full Hash: 4357d695b8d5dd901788d922b73e89f16c45b2c9
# Author: Bob Owen <bobowencode@gmail.com>
# Date: 2019-06-07 16:20:31
# Regressor Bug: 1464032
# File Overlap Count: 1
# Description:
#   Bug 1464032 Part 12: Add CanvasParent, CanvasChild and RecordedTextureData. r=mattwoodrow, jld
#   
#   RecordedTextureData records TextureData calls for play back in the GPU process.
#   CanvasChild and CanvasParent set up the recorder and translator.
#   They also help to manage the starting of translation and co-ordinating the
# ==============================================================================

diff -r e32d6c3ba887 -r 4357d695b8d5 gfx/layers/ipc/CanvasParent.h
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/gfx/layers/ipc/CanvasParent.h	Sun Dec 02 14:19:11 2018 +0000
@@ -0,0 +1,97 @@
+/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set ts=8 sts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef mozilla_layers_CanvasParent_h
+#define mozilla_layers_CanvasParent_h
+
+#include "mozilla/ipc/CrossProcessSemaphore.h"
+#include "mozilla/layers/CanvasTranslator.h"
+#include "mozilla/layers/PCanvasParent.h"
+#include "mozilla/UniquePtr.h"
+
+namespace mozilla {
+namespace layers {
+class TextureData;
+
+class CanvasParent final : public PCanvasParent {
+ public:
+  NS_INLINE_DECL_THREADSAFE_REFCOUNTING(CanvasParent)
+
+  friend class PProtocolParent;
+
+  /**
+   * Create a CanvasParent and bind it to the given endpoint on the
+   * CanvasPlaybackLoop.
+   *
+   * @params aEndpoint the endpoint to bind to
+   * @returns the new CanvasParent
+   */
+  static already_AddRefed<CanvasParent> Create(
+      Endpoint<PCanvasParent>&& aEndpoint);
+
+  static bool IsInCanvasThread();
+
+  /**
+   * Shutdown the canvas thread.
+   */
+  static void Shutdown();
+
+  /**
+   * Create a canvas translator for a particular TextureType, which translates
+   * events from a CanvasEventRingBuffer.
+   *
+   * @param aTextureType the TextureType the translator will create
+   * @param aReadHandle handle to the shared memory for the
+   * CanvasEventRingBuffer
+   * @param aReaderSem reading blocked semaphore for the CanvasEventRingBuffer
+   * @param aWriterSem writing blocked semaphore for the CanvasEventRingBuffer
+   */
+  ipc::IPCResult RecvCreateTranslator(
+      const TextureType& aTextureType,
+      const ipc::SharedMemoryBasic::Handle& aReadHandle,
+      const CrossProcessSemaphoreHandle& aReaderSem,
+      const CrossProcessSemaphoreHandle& aWriterSem);
+
+  /**
+   * Used to tell the CanvasTranslator to start translating again after it has
+   * stopped due to a timeout waiting for events.
+   */
+  ipc::IPCResult RecvResumeTranslation();
+
+  void ActorDestroy(ActorDestroyReason why) final {}
+
+  void DeallocPCanvasParent();
+
+  /**
+   * Used by the compositor thread to get the SurfaceDescriptor associated with
+   * the DrawTarget from another process.
+   *
+   * @param aDrawTarget the key to find the TextureData
+   * @returns the SurfaceDescriptor associated with the key
+   */
+  UniquePtr<SurfaceDescriptor> LookupSurfaceDescriptorForClientDrawTarget(
+      const uintptr_t aDrawTarget);
+
+ private:
+  CanvasParent();
+  ~CanvasParent() final;
+
+  DISALLOW_COPY_AND_ASSIGN(CanvasParent);
+
+  void Bind(Endpoint<PCanvasParent>&& aEndpoint);
+
+  void PostStartTranslationTask();
+
+  void StartTranslation();
+
+  RefPtr<CanvasParent> mSelfRef;
+  UniquePtr<CanvasTranslator> mTranslator;
+};
+
+}  // namespace layers
+}  // namespace mozilla
+
+#endif  // mozilla_layers_CanvasParent_h